### PR TITLE
bike sharing system handling

### PR DESCRIPTION
**Pull request by antoine-de**
Handling bike sharing system (BSS).

To acheive that, some modifications to the street network mechanism have been made:
- prior refactoring of the streetnetork.
  the street network is now delegating all the work to 2 GeoRefPathFinder (one for the departure and one for the arrival)
- changes all distances to durations:
  since several transportation modes can be used (thus several transportation speeds), we consider duration and  not distances in all the algorithm
- we filter the possible transportation modes used in the djisktra.
  - for walking (or biking or car) we can only use the foot network(or bike or car network)
  - for BSS we can use either foot of bike network
- ED Data: 
  The Bicycles stations taken into account are the POI of type 'bicycle_rental' in ED.
  For each station, we create 2 new edges between the walking and the biking graph, one to take the bike, one to give it back.
  The edges are created between the nearest walking edge and the nearest biking edge.
  The duration of the edge is the distance as the crow flies at the default walking speed + the time to either take or hang the bike bake.
- API changes:
  - removal of walking_distance, biking_distance and car_distances parameters.
    Those parameters are replaced by a unique parameter 'max_duration_to_pt' corresponding to the maximum allowed time to get to the public transport and come back from it.
  - for multi transportation results, the street network section is split for each different transportation mean + one section to take the bike and one section to give it back.

### CODE REVIEW

File: source/ed/connectors/osm2nav.cpp [hunk 1]
```diff
@@ -207,15 +211,15 @@ void Visitor::build_vls_edges(){
                 navitia::georef::edge_t e = geo_ref.nearest_edge(poi->coord, u);
                 navitia::georef::Edge edge;
                 edge.way_idx = geo_ref.graph[e].way_idx;
-                edge.length = 0;
-                edge.time = 120; // le temps nécessaire pour prendre le vélo
-                boost::add_edge(u + geo_ref.offsets[navitia::type::Mode_e::Vls], u + geo_ref.offsets[navitia::type::Mode_e::Bike], edge, geo_ref.graph);
-                edge.time = 180; // le temps nécessaire pour déposer le vélo
-                boost::add_edge(u + geo_ref.offsets[navitia::type::Mode_e::Bike], u + geo_ref.offsets[navitia::type::Mode_e::Vls], edge, geo_ref.graph);
+                edge.duration = boost::posix_time::seconds(120); // time needed to take the bike
```
Comment (kinnou02) [inline source/ed/connectors/osm2nav.cpp]: ca vaudrait peut être le coup de mettre les temps en constante dans un coin à coté de navitia::georef::default_speed

File: source/ed/ed_reader.cpp [hunk 2]
```diff
@@ -861,17 +862,21 @@ void EdReader::fill_graph(navitia::type::Data& data, pqxx::work& work){
...
             navitia::georef::Edge e;
-            const_it["leng"].to(e.length);
+            float len;
+            const_it["leng"].to(len);
             e.way_idx = way->idx;
             uint64_t source = this->node_map[const_it["source_node_id"].as<uint64_t>()];
             uint64_t target = this->node_map[const_it["target_node_id"].as<uint64_t>()];
             if ((source != std::numeric_limits<uint64_t>::max()) && (target != std::numeric_limits<uint64_t>::max())){
                 data.geo_ref.ways[way->idx]->edges.push_back(std::make_pair(source, target));
+                e.duration = boost::posix_time::seconds(len / navitia::georef::default_speed[navitia::type::Mode_e::Walking]);
```
Comment (kinnou02) [inline source/ed/ed_reader.cpp:872]: tu as quasiment le meme code dans  Visitor::edges, çà vaudrait le coup de factoriser un peu

File: source/georef/georef.cpp [hunk 3]
```diff
@@ -293,66 +295,83 @@ Path GeoRef::build_path(vertex_t best_destination, std::vector<vertex_t> preds,
...
-    } else {
-        throw proximitylist::NotFound();
+    current = preds[best_destination]; //we skip the middle point since it has already been added
+    while (current != preds[current]) {
+        reverse_path.push_back(current);
+        current = preds[current];
     }
+    reverse_path.push_back(current);
+
+//    std::cout << "to go to " << best_destination << "[" << (int)(get_mode(best_destination)) << "] we pass through :" << std::endl;
```
Comment (kinnou02) [inline source/georef/georef.cpp]: ca ressemble à du code de debug çà

File: source/georef/georef.cpp [hunk 4]
```diff
@@ -293,66 +295,83 @@ Path GeoRef::build_path(vertex_t best_destination, std::vector<vertex_t> preds,
...
+//        std::cout << v << "[" << (int)(get_mode(v)) << "] " << std::endl;
+//    }
 
+    return build_path(reverse_path, false);
 }
 
 
+void GeoRef::add_way(const Way& w){
+    Way* to_add = new Way;
+    to_add->name =w.name;
```
Comment (kinnou02) [inline source/georef/georef.cpp]: l'espace! :@

File: source/georef/georef.h [hunk 5]
```diff
@@ -15,14 +15,24 @@
...
+#include "utils/exception.h"
 #include "utils/flat_enum_map.h"
 
-
 namespace nt = navitia::type;
 namespace nf = navitia::autocomplete;
 
 namespace navitia { namespace georef {
 
+/// default speed by transportation mode, defined at compile time
```
Comment (kinnou02) [inline source/georef/georef.h]: en carottes?

File: source/routing/raptor_init.cpp [hunk 6]
```diff
@@ -254,17 +258,19 @@ bool improves(const DateTime & best_so_far, bool clockwise, const DateTime & cur
...
+                     bool clockwise, const std::vector<label_vector_t> &labels, const type::Data &data) {
 
     const type::JourneyPatternPoint* current_jpp = data.pt_data.journey_pattern_points[jpp_idx];
     int cnt = count;
-    float walking_time = 0;
+    bt::time_duration walking_time = {};
 
     //Marche à la fin
-    for(auto dest_dist : destinations) {
+    for(auto dest_dist : destinations) {//TODO remplacer par un find
```
Comment (kinnou02) [inline source/routing/raptor_init.cpp]: tu as oublié le todo

Comment (antoine-de) [inline source/routing/raptor_init.cpp]: en fait je calerais un find quand on aura des lambda polymorphic, la c'est plus long a ecrire :p

**Comment by antoine-de** [comment]
The branch has been merged in dev in an 'usual' way (who said blunder ?)

So the pull request can be closed without been merged.

**Pull request closed without merging**